File: ios/Views/MotionBlurView.swift
//
//  MotionBlurView.swift
//  Pods
//
//  Created by rit3zh CX on 9/23/26.
//

import ExpoModulesCore
import UIKit

final class MotionBlurView: ExpoView, FrameDriverClient {
  private static let showThreshold: CGFloat = 1.0
  private static let hideThreshold: CGFloat = 0.5
  private static let resourceIdleTimeout: CFTimeInterval = 1.5
  private static let maxInkOverflow: CGFloat = 96
  // A spring that bounces stops for a moment at each peak. Motion that resumes within this window
  // reuses the snapshot instead of capturing the content again mid-animation.
  private static let snapshotReuseWindow: CFTimeInterval = 0.3

  var configuration = MotionBlurConfiguration() {
    didSet {
      configurationDidChange(from: oldValue)
    }
  }

  private let contentView = MotionBlurContentView()
  private let tracker = MotionTracker()
  private var overlayView: MotionBlurOverlayView?
  private var renderer: MotionBlurRenderer?
  private var canvas: BlurCanvas?
  private var isBlurVisible = false
  private var needsSnapshot = true
  private var isRegistered = false
  private var lastActiveTimestamp: CFTimeInterval = 0
  private var blurHiddenTimestamp: CFTimeInterval = -.infinity
  required init(appContext: AppContext? = nil) {
    super.init(appContext: appContext)
    clipsToBounds = false
    contentView.frame = bounds
    contentView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
    addSubview(contentView)
  }

  @objc var effectiveContentView: UIView {
    contentView
  }

  override var bounds: CGRect {
    didSet {
      if bounds != oldValue {
        contentView.frame = bounds
      }
    }
  }

  override func mountChildComponentView(_ childComponentView: UIView, index: Int) {
    super.mountChildComponentView(childComponentView, index: index)
    needsSnapshot = true
  }

  override func unmountChildComponentView(_ childComponentView: UIView, index: Int) {
    super.unmountChildComponentView(childComponentView, index: index)
    needsSnapshot = true
  }

  func propsDidUpdate() {
    needsSnapshot = true
    updateCanvas()
    updateShadowPath()
  }

  override func layoutSubviews() {
    super.layoutSubviews()
    contentView.frame = bounds
    updateCanvas()
    updateShadowPath()
  }

  override func didMoveToWindow() {
    super.didMoveToWindow()
    updateCanvas()
    updateRegistration()
  }

  func frameDriver(_ driver: FrameDriver, didTickAt timestamp: CFTimeInterval, batch: RenderBatch) -> Bool {
    guard let window, let frame = currentWindowFrame(in: window) else {
      tracker.reset()
      setBlurVisible(false)
      return false
    }
    tracker.update(frame: frame, bounds: bounds, timestamp: timestamp)

    let halfStreak = min(tracker.velocity.maxSpeed(in: bounds) * configuration.exposure, configuration.intensity)
    let halfStreakPixels = halfStreak * (canvas?.scale ?? 1)
    let threshold = isBlurVisible ? Self.hideThreshold : Self.showThreshold
    let wantsBlur = halfStreakPixels > threshold && isOnScreen(frame, in: window)
    setBlurVisible(wantsBlur && drawBlur(halfStreakPixels: halfStreakPixels, into: batch))

    let isActive = tracker.isMoving || isBlurVisible
    if isActive {
      lastActiveTimestamp = timestamp
    } else if timestamp - lastActiveTimestamp > Self.resourceIdleTimeout {
      releaseGPUResources()
    }
    return isActive
  }

  func frameDriverDidReceiveMemoryWarning(_ driver: FrameDriver) {
    releaseGPUResources()
  }

  private func drawBlur(halfStreakPixels: CGFloat, into batch: RenderBatch) -> Bool {
    guard !isHidden, alpha > 0.01, let metal = MetalContext.shared, metal.pipelineState != nil else {
      return false
    }
    let renderer = renderer ?? MotionBlurRenderer(metal: metal)
    self.renderer = renderer
    let isResumingMotion = !isBlurVisible && CACurrentMediaTime() - blurHiddenTimestamp < Self.snapshotReuseWindow
    let needsCapture = needsSnapshot || configuration.isLive || (!isBlurVisible && !isResumingMotion)
    if needsCapture {
      updateCanvas()
    }
    guard let canvas, canvas.isRenderable else {
      return false
    }
    let overlay = overlayView ?? makeOverlay(device: metal.device, canvas: canvas)

    if needsCapture || !renderer.hasSnapshot(for: canvas) {
      guard captureContent(with: renderer, canvas: canvas) else {
        return false
      }
      needsSnapshot = false
    }
    return renderer.draw(
      velocity: tracker.velocity,
      halfStreakPixels: halfStreakPixels,
      configuration: configuration,
      canvas: canvas,
      into: overlay.metalLayer,
      batch: batch
    )
  }

  private func captureContent(with renderer: MotionBlurRenderer, canvas: BlurCanvas) -> Bool {
    let isContentHidden = contentView.alpha < 1
    // Glass and blur effects only show up in an on-screen snapshot, which needs the content visible.
    let drawsHierarchy = !isContentHidden && Self.containsVisualEffect(contentView)
    if isContentHidden {
      contentView.alpha = 1
    }
    defer {
      if isContentHidden {
        contentView.alpha = 0
      }
    }
    return renderer.captureSnapshot(of: contentView, canvas: canvas, drawsHierarchy: drawsHierarchy)
  }

  private static func containsVisualEffect(_ view: UIView) -> Bool {
    if view is UIVisualEffectView {
      return true
    }
    return view.subviews.contains { !$0.isHidden && containsVisualEffect($0) }
  }

  private func setBlurVisible(_ visible: Bool) {
    guard visible != isBlurVisible else {
      return
    }
    isBlurVisible = visible
    if !visible {
      blurHiddenTimestamp = CACurrentMediaTime()
    }
    contentView.alpha = visible ? 0 : 1
    overlayView?.isHidden = !visible
    updateShadowPath()
  }

  private func makeOverlay(device: MTLDevice, canvas: BlurCanvas) -> MotionBlurOverlayView {
    let overlay = MotionBlurOverlayView(device: device)
    overlay.apply(canvas)
    addSubview(overlay)
    overlayView = overlay
    return overlay
  }

  private func releaseGPUResources() {
    guard !isBlurVisible else {
      return
    }
    renderer = nil
    overlayView?.removeFromSuperview()
    overlayView = nil
    needsSnapshot = true
  }

  private func updateShadowPath() {
    let contentLayer = contentView.layer
    guard
      layer.shadowOpacity > 0, !isBlurVisible,
      let backgroundColor = contentLayer.backgroundColor, backgroundColor.alpha > 0.999,
      bounds.width > 0, bounds.height > 0
    else {
      layer.shadowPath = nil
      return
    }
    let radius = min(contentLayer.cornerRadius, bounds.width / 2, bounds.height / 2)
    layer.shadowPath = CGPath(roundedRect: bounds, cornerWidth: radius, cornerHeight: radius, transform: nil)
  }

  private func currentWindowFrame(in window: UIWindow) -> AffineMap? {
    let source: CALayer
    let destination: CALayer
    if let presentedLayer = layer.presentation(), let presentedWindowLayer = window.layer.presentation() {
      source = presentedLayer
      destination = presentedWindowLayer
    } else {
      source = layer
      destination = window.layer
    }
    let rect = source.bounds
    return AffineMap(
      mapping: rect,
      topLeft: source.convert(CGPoint(x: rect.minX, y: rect.minY), to: destination),
      topRight: source.convert(CGPoint(x: rect.maxX, y: rect.minY), to: destination),
      bottomLeft: source.convert(CGPoint(x: rect.minX, y: rect.maxY), to: destination)
    )
  }

  private func isOnScreen(_ frame: AffineMap, in window: UIWindow) -> Bool {
    guard let canvas else {
      return false
    }
    let rect = canvas.frame
    let corners = [
      frame.apply(to: CGPoint(x: rect.minX, y: rect.minY)),
      frame.apply(to: CGPoint(x: rect.maxX, y: rect.minY)),
      frame.apply(to: CGPoint(x: rect.minX, y: rect.maxY)),
      frame.apply(to: CGPoint(x: rect.maxX, y: rect.maxY)),
    ]
    let xs = corners.map(\.x)
    let ys = corners.map(\.y)
    let boundingBox = CGRect(
      x: xs.min() ?? 0,
      y: ys.min() ?? 0,
      width: (xs.max() ?? 0) - (xs.min() ?? 0),
      height: (ys.max() ?? 0) - (ys.min() ?? 0)
    )
    return boundingBox.intersects(window.bounds)
  }

  private func contentInkRect() -> CGRect {
    let contentLayer = contentView.layer
    var ink = contentView.bounds
    for sublayer in contentLayer.sublayers ?? [] where !sublayer.isHidden && sublayer.opacity > 0 {
      ink = ink.union(sublayer.frame)
      if LayerTreeRenderer.hasVisibleShadow(sublayer) {
        let caster = sublayer.shadowPath?.boundingBoxOfPath ?? sublayer.bounds
        let spread = sublayer.shadowRadius * 3
        let shadow = sublayer.convert(caster, to: contentLayer)
          .offsetBy(dx: sublayer.shadowOffset.width, dy: sublayer.shadowOffset.height)
          .insetBy(dx: -spread, dy: -spread)
        ink = ink.union(shadow)
      }
    }
    let inkInHost = ink.offsetBy(dx: bounds.minX, dy: bounds.minY)
    return inkInHost.intersection(bounds.insetBy(dx: -Self.maxInkOverflow, dy: -Self.maxInkOverflow))
  }

  private func updateCanvas() {
    guard let window else {
      return
    }
    let newCanvas = BlurCanvas(
      bounds: bounds,
      inkRect: contentInkRect(),
      intensity: configuration.intensity,
      scale: window.screen.scale
    )
    guard newCanvas != canvas else {
      return
    }
    canvas = newCanvas
    needsSnapshot = true
    overlayView?.apply(newCanvas)
  }

  private func configurationDidChange(from oldValue: MotionBlurConfiguration) {
    if configuration.intensity != oldValue.intensity {
      updateCanvas()
    }
    if configuration.isEnabled != oldValue.isEnabled {
      updateRegistration()
    }
  }

  private func updateRegistration() {
    let shouldRun = window != nil && configuration.isEnabled && MetalContext.shared != nil
    guard shouldRun != isRegistered else {
      return
    }
    isRegistered = shouldRun

    if shouldRun {
      MetalContext.shared?.prepare()
      FrameDriver.shared.add(self)
    } else {
      FrameDriver.shared.remove(self)
      tracker.reset()
      setBlurVisible(false)
      releaseGPUResources()
    }
  }
}

File: android/src/main/java/expo/modules/motionblurview/views/MotionBlurView.kt
package expo.modules.motionblurview.views

import android.content.Context
import android.graphics.Canvas
import android.graphics.Matrix
import android.graphics.RecordingCanvas
import android.graphics.RectF
import android.os.Build
import android.util.Log
import android.view.View
import android.view.ViewTreeObserver
import android.view.animation.AnimationUtils
import androidx.annotation.RequiresApi
import expo.modules.kotlin.AppContext
import expo.modules.kotlin.views.ExpoView
import expo.modules.motionblurview.models.AffineMap
import expo.modules.motionblurview.models.BlurCanvas
import expo.modules.motionblurview.models.Bounds
import expo.modules.motionblurview.models.MotionBlurConfiguration
import expo.modules.motionblurview.models.Point
import expo.modules.motionblurview.motion.MotionTracker
import expo.modules.motionblurview.rendering.MotionBlurRenderer
import kotlin.math.max
import kotlin.math.min

class MotionBlurView(context: Context, appContext: AppContext) :
  ExpoView(context, appContext), ViewTreeObserver.OnPreDrawListener {

  internal var configuration = MotionBlurConfiguration.DEFAULT
    set(value) {
      val oldValue = field
      field = value
      configurationDidChange(oldValue)
    }

  private val tracker = MotionTracker()
  private val globalMatrix = Matrix()
  private val points = FloatArray(6)
  private val childRect = RectF()
  private var renderer: MotionBlurRenderer? = null
  private var blurCanvas: BlurCanvas? = null
  private var halfStreakPixels = 0.0
  private var isBlurVisible = false
  private var registeredObserver: ViewTreeObserver? = null

  private val density: Double
    get() = resources.displayMetrics.density.toDouble()

  init {
    clipChildren = false
    clipToPadding = false
    setWillNotDraw(false)
  }

  override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
    setMeasuredDimension(MeasureSpec.getSize(widthMeasureSpec), MeasureSpec.getSize(heightMeasureSpec))
  }

  override fun onLayout(changed: Boolean, l: Int, t: Int, r: Int, b: Int) = Unit

  override fun onAttachedToWindow() {
    super.onAttachedToWindow()
    updateRegistration(isAttached = true)
  }

  override fun onDetachedFromWindow() {
    updateRegistration(isAttached = false)
    super.onDetachedFromWindow()
  }

  fun propsDidUpdate() {
    invalidate()
  }

  override fun onPreDraw(): Boolean {
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
      tick()
    }
    return true
  }

  override fun draw(canvas: Canvas) {
    if (isBlurVisible && Build.VERSION.SDK_INT >= Build.VERSION_CODES.S && drawBlur(canvas)) {
      return
    }
    super.draw(canvas)
  }

  @RequiresApi(Build.VERSION_CODES.S)
  private fun tick() {
    val frame = if (isShown) currentWindowFrame() else null
    if (frame == null) {
      tracker.reset()
      setBlurVisible(false)
      return
    }
    val bounds = contentBounds()
    val timestamp = AnimationUtils.currentAnimationTimeMillis() / 1000.0
    tracker.update(frame, bounds, timestamp)

    val halfStreak = min(tracker.velocity.maxSpeed(bounds) * configuration.exposure, configuration.intensity)
    halfStreakPixels = halfStreak * density
    val threshold = if (isBlurVisible) HIDE_THRESHOLD else SHOW_THRESHOLD
    val canvas = updateCanvas()
    val wantsBlur = halfStreakPixels > threshold && alpha > 0.01f && canvas != null && isOnScreen(frame, canvas)
    setBlurVisible(wantsBlur)

    if (isBlurVisible) {
      invalidate()
    }
    if (isBlurVisible || tracker.isMoving) {
      postInvalidateOnAnimation()
    }
  }

  @RequiresApi(Build.VERSION_CODES.S)
  private fun drawBlur(target: Canvas): Boolean {
    val recording = target as? RecordingCanvas ?: return false
    val canvas = blurCanvas?.takeIf { it.isRenderable } ?: return false
    val renderer = renderer ?: MotionBlurRenderer.create()?.also { renderer = it } ?: return false
    return renderer.draw(recording, canvas, tracker.velocity, halfStreakPixels, configuration) { contentCanvas ->
      super.draw(contentCanvas)
    }
  }

  private fun setBlurVisible(visible: Boolean) {
    if (visible == isBlurVisible) {
      return
    }
    isBlurVisible = visible
    if (!visible) {
      renderer?.release()
    }
    invalidate()
  }

  private fun contentBounds(): Bounds = Bounds(0.0, 0.0, width / density, height / density)

  @RequiresApi(Build.VERSION_CODES.Q)
  private fun currentWindowFrame(): AffineMap? {
    if (!isAttachedToWindow || width <= 0 || height <= 0) {
      return null
    }
    globalMatrix.reset()
    transformMatrixToGlobal(globalMatrix)
    points[0] = 0f
    points[1] = 0f
    points[2] = width.toFloat()
    points[3] = 0f
    points[4] = 0f
    points[5] = height.toFloat()
    globalMatrix.mapPoints(points)
    val density = density
    return AffineMap.mapping(
      contentBounds(),
      topLeft = Point(points[0] / density, points[1] / density),
      topRight = Point(points[2] / density, points[3] / density),
      bottomLeft = Point(points[4] / density, points[5] / density)
    )
  }

  @RequiresApi(Build.VERSION_CODES.Q)
  private fun isOnScreen(frame: AffineMap, canvas: BlurCanvas): Boolean {
    val root = rootView
    globalMatrix.reset()
    root.transformMatrixToGlobal(globalMatrix)
    points[0] = 0f
    points[1] = 0f
    points[2] = root.width.toFloat()
    points[3] = root.height.toFloat()
    globalMatrix.mapPoints(points, 0, points, 0, 2)
    val density = density
    val window = Bounds(points[0] / density, points[1] / density, points[2] / density, points[3] / density)
    return Bounds.enclosing(canvas.frame.corners.map(frame::apply)).intersects(window)
  }

  private fun contentInkRect(bounds: Bounds): Bounds {
    val density = density
    var ink = bounds
    for (index in 0 until childCount) {
      val child = getChildAt(index)
      if (child.visibility != View.VISIBLE || child.alpha <= 0f) {
        continue
      }
      childRect.set(0f, 0f, child.width.toFloat(), child.height.toFloat())
      child.matrix.mapRect(childRect)
      childRect.offset(child.left.toFloat(), child.top.toFloat())
      val shadow = max(child.z, 0f) * SHADOW_SPREAD_PER_ELEVATION
      childRect.inset(-shadow, -shadow)
      ink = ink.union(
        Bounds(
          childRect.left / density,
          childRect.top / density,
          childRect.right / density,
          childRect.bottom / density
        )
      )
    }
    return ink.intersection(bounds.insetBy(-MAX_INK_OVERFLOW, -MAX_INK_OVERFLOW))
  }

  private fun updateCanvas(): BlurCanvas? {
    val bounds = contentBounds()
    val canvas = BlurCanvas.make(
      bounds = bounds,
      inkRect = contentInkRect(bounds),
      intensity = configuration.intensity,
      scale = density,
      streakReach = MotionBlurRenderer.streakReach
    )
    blurCanvas = canvas
    return canvas.takeIf { it.isRenderable }
  }

  private fun configurationDidChange(oldValue: MotionBlurConfiguration) {
    if (configuration.isEnabled != oldValue.isEnabled) {
      updateRegistration(isAttachedToWindow)
    }
  }

  private fun updateRegistration(isAttached: Boolean) {
    val shouldRun = isAttached && configuration.isEnabled && MotionBlurRenderer.isSupported
    if (shouldRun == (registeredObserver != null)) {
      return
    }
    if (shouldRun) {
      val observer = viewTreeObserver
      observer.addOnPreDrawListener(this)
      registeredObserver = observer
      if (renderer == null) {
        renderer = MotionBlurRenderer.create()
      }
    } else {
      val observer = registeredObserver?.takeIf { it.isAlive } ?: viewTreeObserver
      observer.removeOnPreDrawListener(this)
      registeredObserver = null
      tracker.reset()
      setBlurVisible(false)
      renderer?.release()
      renderer = null
    }
    if (isAttached && configuration.isEnabled && !MotionBlurRenderer.isSupported && !didWarnUnsupported) {
      didWarnUnsupported = true
      Log.i(TAG, "Motion blur requires Android 12 (API 31). The children render without blur on API ${Build.VERSION.SDK_INT}.")
    }
  }

  private companion object {
    const val TAG = "ExpoMotionBlur"
    const val SHOW_THRESHOLD = 1.0
    const val HIDE_THRESHOLD = 0.5
    const val MAX_INK_OVERFLOW = 96.0
    const val SHADOW_SPREAD_PER_ELEVATION = 2f
    var didWarnUnsupported = false
  }
}
